docs/plugins/asg_rollout.md

```diff
@@ -0,0 +1,85 @@
+# Auto-Scaling Group LaunchConfig Rollout Plugin
+
+## Overview
+
+This plugin adds support for rolling out changes to Auto Scaling
+Groups that have happened after a stack update. It supports various
+pre- and post-actions on the instances (using Moonshot's native SSH
+support).
+
+## Example
+
+The ASGRollout class is intended to be used within a method in your
+project's derived class of the Moonshot::CLI class.
+
+```ruby
+#!/usr/bin/env ruby
+
+require 'moonshot'
+
+class MyService < Moonshot::CLI
+  # .. normal configuration ..
+  
+  desc :asg_rollout, 'Update instances in the Auto Scaling Group'
+  def asg_rollout
+    Moonshot::Plugins::ASGRollout.new(
+      controller: controller,
+      logical_id: 'APIAutoScalingGroup',
+      # Trigger the worker process to shut down cleanly.
+      pre_detach: -> {
+        instance.exec('systemctl stop my-worker-1')
+      }
+
+      # Ensure that the worker process has completed it's work before terminating.
+      terminate_when: -> {
+        0 == instance.exec('pgrep -f my-worker-1').exitstatus
+      }
+    ).run!
+  end
+end
+
+MyService.start
+```
+
+## Configuration
+
+The ASGRollout object accepts two required keyword arguments:
+  - **controller**: A Moonshot::Controller. If you are using
+    Moonshot::CLI the `controller` method is most often when you want
+    to use.
+  - **logical_id**: The Logical resource ID from the CloudFormation
+    stack to operate on.
+
+By default, the ASGRollout class will build a list of instances with
+non-conforming LaunchConfiguration, then one at a time perform the
+following actions:
+  1. Increase the *Max* and *Desired* sizes of the Auto Scaling Group by one.
+  1. Wait for the new instance to be *InService*.
+  1. Detach a non-conforming instance.
+  1. TODO: Do we need to wait for connection draining or is that
+     handled by the ELB detachment mechiansm?
+  1. Wait the duration of the connecti
+  1. Wait for a new instance to replace it.
+  1. Wait for that new instance to be *InService*.
+  1. Terminate the detached non-conforming instance.
```

Comment:
What does this mean? In our case it's `ssh shutdown`. Does it need to be a lambda as well?